pandas/core/reshape/concat.py

```diff
@@ -611,8 +630,10 @@ def get_result(self):
             new_data = concatenate_managers(
                 mgrs_indexers, self.new_axes, concat_axis=self.bm_axis, copy=self.copy
             )
-            if not self.copy:
+            if not self.copy and not _using_copy_on_write():
                 new_data._consolidate_inplace()
+            elif _using_copy_on_write() and not self.copy:
+                new_data = new_data.copy(deep=False)
```

Comment:
I would expect that `concatenate_managers` already takes care of setting up references. Of course, `concatenate_managers` is only used here, so it if it easier to implement this way, that's fine as well (but maybe add a comment about it that `concatenate_managers` basically returned a view and didn't yet set up `new_data.refs`)

Reply:
It _might_ also not be sufficient this way for keeping track of the reference from parent to child. Because `new_data` here will reference this intermediate BlockManager, but not the original BlockManager from the concat input. So the manager of the parent dataframes are not referenced by any result.

Reply:
Checking out this branch, I can confirm that is indeed an issue with the current implementation:

```
In [1]: pd.options.mode.copy_on_write = True

In [2]: df = DataFrame({"b": ["a"] * 3})
   ...: df2 = DataFrame({"a": ["a"] * 3})

In [3]: result = pd.concat([df, df2], axis=1)

In [4]: result
Out[4]: 
   b  a
0  a  a
1  a  a
2  a  a

In [8]: df._mgr._has_no_reference(0)
Out[8]: True

In [9]: df.iloc[0, 0] = 'c'

In [10]: result
Out[10]: 
   b  a
0  c  a
1  a  a
2  a  a
```

Reply:
Pushed the logic down and added your case as test. Works now